Make revocation of access tokens optional

NobodysNightmare opened the pull request:
### Summary

As discussed in #1314, the automatic revocation of old access tokens can lead to problems when multiple physical machines (clients) share the same set of credentials (e.g. multiple web servers behind a load balancer).

However, the revocation was implemented, to make the fetching of tokens more efficient, since it did not perform well for large amounts of tokens.

Therefore this pull request implements two changes:
* filtering by the TTL of a token when fetching tokens from the database (thus also reducing the number of fetched tokens)
* making the automatic revocation introduced in #1271 optional (disabled by default)

### Considerations

I decided to disable automatic revocation by default. My reasoning:
* It was a breaking change introduced in 5.2, so it was a surprise for anyone upgrading from 5.1 and below
* I am not sure if there is real _demand_ for that feature or whether it was purely out of performance considerations... But I can at least imagine the demand

Fetching based on TTL is not working when you set a `custom_access_token_expires_in`, since we could not guarantee that the tokens being filtered out would be considered as expired by the custom strategy. I am pretty sad about introducing this piece of "works different depending on configuration" and I am not sure whether we should at least point it out in the documentation.

NobodysNightmare commented on lib/doorkeeper/oauth/client_credentials/creator.rb:
```diff
@@ -11,7 +11,9 @@ def call(client, scopes, attributes = {})
             return existing_token
```
Note: We are still always fetching an `existing_token`, even though it is perfectly possible, that `reuse_access_token` and `revoke_previous_client_credentials_token` are both disabled. In that case we would not really need the existing token and fetching it is just slowing us down.

I'd prefer to "hide" it behind a memoizing method, but since we need arguments from the call to `call` to obtain the token, this seems unfeasible.

What would you think about fetching it like this:

```ruby
existing_token = existing_token_for(...) if needs_existing_token?

def needs_existing_token?
  Doorkeeper.configuration.reuse_access_token || Doorkeeper.configuration.revoke_previous_client_credentials_token
end
```

nbulaj replied:
Sounds reasonable to me

NobodysNightmare replied:
done

NobodysNightmare commented:
In #1314 I proposed that we also adapt the changelog for 5.2, to note that the automatic revocation might be an **[IMPORTANT]** change.

Should we do this? in this PR?

NobodysNightmare commented:
> 	Please squash all your commits to a single one

Is this desirable here? I feel that the changes belong together (one PR), but are distinct enough to justify documenting them in different commits.

Should I split this pull request into two requests?

nbulaj commented:
> It was a breaking change introduced in 5.2, so it was a surprise for anyone upgrading from 5.1 and below

Actually agree. 

> I proposed that we also adapt the changelog for 5.2, to note that the automatic revocation might be an [IMPORTANT] change.

Also agree. Could you please add it? Also it would be great to add an entry to the wiki in Upgrading section.

> Is this desirable here? I feel that the changes belong together (one PR), but are distinct enough to justify documenting them in different commits.

I can do a squash merge after we finish with this PR so you can don't mind about it. Btw, there is no reason to have a full history for any particular PR - we have everything discussed in the PR / commit description / etc.

nbulaj commented on lib/doorkeeper/models/access_token_mixin.rb:
```diff
@@ -191,10 +191,23 @@ def find_or_create_for(application, resource_owner_id, scopes, expires_in, use_r
                  revoked_at: nil)
+
+        if ttl
+          tokens.where("created_at >= ?", ttl)
```
Can we move it to the scope? We need it for any other ORM extension to make integrration as easy as possible. So mongoid and sequel extensions can just introduce their own scopes instead of reimplementing this method fully.

NobodysNightmare replied:
I now added a scope `Doorkeeper::AccessToken.created_since(time)`, I hope that was what you were asking for ^^

nbulaj commented on lib/doorkeeper/models/access_token_mixin.rb:
```diff
@@ -191,10 +191,23 @@ def find_or_create_for(application, resource_owner_id, scopes, expires_in, use_r
+        # In case of static token expiry times, we can optimize fetching tokens
+        # by already filtering out tokens whose TTL has expired.
+        # We can't do that if custom expirations times have been configured, though.
+        ttl = unless Doorkeeper.configuration.option_defined?(:custom_access_token_expires_in)
```
This construction looks complex to me :thinking: assignment with `unless`... Is it possible to simplify it in order to make more human-friendly aka readable?

NobodysNightmare replied:
replaced it with assignment inside the unless

nbulaj commented on lib/doorkeeper/oauth/client_credentials/creator.rb:
```diff
@@ -11,7 +11,9 @@ def call(client, scopes, attributes = {})
           end
 
-          existing_token&.revoke
+          if Doorkeeper.configuration.revoke_previous_client_credentials_token
```
Could we add a new private method called `configuration` as a synthetic sugar for accessing `Doorkeeper.configuration` ? Maybe a delegator I don't know. I see we call it multiple times for this module, we can simplify some constructions.

NobodysNightmare replied:
done

NobodysNightmare commented on CHANGELOG.md:
```diff
@@ -38,6 +46,11 @@ User-visible changes worth mentioning.
+
+**[IMPORTANT]** If you rely on being able to fetch multiple access tokens from the same
+client using client credentials flow, you should skip to version 5.3, where this behaviour
+is deactivated by default.
```
Now that I wrote this message I realized that I didn't exactly know _what_ I want to point out here, please have a look:

* should I already refer to this as 5.3? Even though it is not yet released and I don't know if it will be called 5.3
* Does the hint make sense at all? I feel this is the best hint I could give, but I had a bad gut feeling writing it.

nbulaj replied:
I think it would be 5.2.2

What about the hint... it's hard to say actually...

nbulaj commented:
Let's summarize if this PR breaks anything within previous releases and do we need to updagrate Wiki guides ?

Sorry I'm too busy and can't currently re-check everything, so any help would be appreciated :handshake:

felipeelias commented on lib/generators/doorkeeper/templates/initializer.rb:
```diff
@@ -98,6 +98,16 @@
+  #
+  # When enabling this option, make sure that you do not expect multiple processes
+  # using the same credentials at the same time (e.g. web servers spanning
+  # multiple machines and/or processes).
```
I'm confused by this comment. Aren't most of the clients nowadays multi-thread/process/machine?

NobodysNightmare replied:
I would say yes. On the other hand I can imagine doorkeeper being deployed in small-scale setups, where you probably don't even think about redundancy. E.g. if I was using doorkeeper for my home projects, that would probably be on a single raspberry-pi or vServer and not in a Kubernetes cluster...

Full disclosure: I don't quite understand the use case of the feature there. I just realized that it was an unexpected breaking change for a multi-server setup. Therefore I made it optional and added the hint to not enable it for multi-machine setups.

hugopeixoto commented on lib/doorkeeper/models/access_token_mixin.rb:
```diff
@@ -191,10 +191,23 @@ def find_or_create_for(application, resource_owner_id, scopes, expires_in, use_r
+          valid_from = Doorkeeper.configuration.access_token_expires_in.seconds.ago
+        end
+
+        tokens = ordered_by(:created_at, :desc)
```
As mentioned in https://github.com/doorkeeper-gem/doorkeeper/issues/1314#issuecomment-540625227, I think using time math would handle every scenario.

We have `expires_in` stored in the database, so something like this could work:

```
tokens
  .where(
    application_id: application_id,
    resource_owner_id: resource_owner_id,
    revoked_at: nil,
  )
  .where("created_at + expires_in * interval '1 second' > ?", Time.now.utc)
```

I'm not sure if this syntax covers every supported database.

nbulaj replied:
> I'm not sure if this syntax covers every supported database.

We must support all the dbs that are supported by ActiveRecord. So either implement such query our self in the adapters , either .. I don't know..

hugopeixoto replied:
Double checked. This syntax won't work in sqlite3. Unsure how to do this. Creating the `expires_at` column would sidestep this whole issue, but I see why we'd want to avoid the redundancy. I'm guessing we're storing `expires_in` because that's the recommended way of reporting the expiration in oauth (https://tools.ietf.org/html/rfc6749#section-4.2.2) but maybe `expires_in` should be calculated from `expires_at`, instead of the other way round? Maybe this has been discussed before.

EugeneFractal replied:
Something like this may work (have not tested):

```ruby
tokens
  .where(
    application_id: application_id,
    resource_owner_id: resource_owner_id,
    revoked_at: nil,
  )
  .where(
    if ActiveRecord::Base.connection.adapter_name == "SQLite"
      "datetime(created_at, '+' || expires_in || ' seconds') > ?"
    else
      "created_at + expires_in * interval '1 second' > ?"
    end,
    Time.now.utc,
  )
```
Thoughts?

nbulaj replied:
@EugeneFractal

It would work, yep. But I'm afraid of timezones used by databases (Postgres, Mysql, etc). Any project could set their own local timezones both for project and DBMS so we could have invalid queries. We can try to convert expressions like `created_at + expires_in * interval '1 second'` to UTC in the database, but every database has it's own syntax for this.

nbulaj replied:
Something generic but very dangerous would be:

```ruby
def expired_for(application_id, resource_owner_id)
  where.not(expires_at: nil)
    .where(
      application_id: application_id,
      resource_owner_id: resource_owner_id,
      revoked_at: nil,
    )
    .where("#{expiration_date_sql_expression} <= ?", Time.now.utc)
end

def expiration_date_sql_expression
  case ActiveRecord::Base.connection.adapter_name.downcase
  when "sqlite", "sqlite3"
    Arel.sql("DATETIME(created_at, '+' || expires_in || ' SECONDS')")
  when "postgis", "postgresql"
    Arel.sql("created_at + expires_in * INTERVAL '1 SECOND' AT TIME ZONE 'UTC'")
  when "mysql", "mysql2"
    Arel.sql("created_at + expires_in * INTERVAL '1 SECOND'")
  when "sqlserver"
    Arel.sql("DATEADD(second, expires_in, created_at) AT TIME ZONE 'UTC'")
  when "oracleenhanced"
    Arel.sql("created_at + INTERVAL to_char(expires_in) second")
  else
    raise "Unsupported database!"
  end
end
```

NobodysNightmare commented:
Remember the custom expiry setting that accepts a block. IIRC this stopped
me from going completely on time comparison.

Nikita Bulai <notifications@github.com> schrieb am Mi., 11. Dez. 2019,
14:05:

> *@nbulaj* commented on this pull request.
> ------------------------------
>
> In lib/doorkeeper/models/access_token_mixin.rb
> <https://github.com/doorkeeper-gem/doorkeeper/pull/1318#discussion_r356586138>
> :
>
> > @@ -191,10 +191,23 @@ def find_or_create_for(application, resource_owner_id, scopes, expires_in, use_r
>        # @return [Doorkeeper::AccessToken] array of matching AccessToken objects
>        #
>        def authorized_tokens_for(application_id, resource_owner_id)
> -        ordered_by(:created_at, :desc)
> +        # In case of static token expiry times, we can optimize fetching tokens
> +        # by already filtering out tokens whose TTL has expired.
> +        # We can't do that if custom expirations times have been configured, though.
> +        unless Doorkeeper.configuration.option_defined?(:custom_access_token_expires_in)
> +          valid_from = Doorkeeper.configuration.access_token_expires_in.seconds.ago
> +        end
> +
> +        tokens = ordered_by(:created_at, :desc)
>
> I'm not sure if this syntax covers every supported database.
>
> We must support all the dbs that are supported by ActiveRecord. So either
> implement such query our self in the adapters , either .. I don't know..
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/doorkeeper-gem/doorkeeper/pull/1318?email_source=notifications&email_token=AADOXUBULQHCOXDBSDDIV3DQYDQQJA5CNFSM4JBVJ6I2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCOZWHUI#discussion_r356586138>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADOXUFSALY66L6QRJOD3IDQYDQQJANCNFSM4JBVJ6IQ>
> .
>

hugopeixoto commented:
> Remember the custom expiry setting that accepts a block. IIRC this stopped me from going completely on time comparison.

I think the block is used to calculate the expiry date upon token creation, and it gets stored in `expires_in`. I don't think it's used every time we need to check the expiration.

Koen03 commented:
Is there any progress in this PR?

NobodysNightmare commented:
Whoops. I was thinking this already got merged..apparently not.

Anything I need to do?

koensengers <notifications@github.com> schrieb am Di., 21. Jan. 2020, 11:34:

> Is there any progress in this PR?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/doorkeeper-gem/doorkeeper/pull/1318?email_source=notifications&email_token=AADOXUDEPINIBRSUIOLSXV3Q63FU3A5CNFSM4JBVJ6I2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJPIQZQ#issuecomment-576620646>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADOXUHGE7QI4XL335YSUP3Q63FU3ANCNFSM4JBVJ6IQ>
> .
>

Koen03 commented:
merge master i think and thats it

Koen03 commented:
> Whoops. I was thinking this already got merged..apparently not. Anything I need to do? koensengers <notifications@github.com> schrieb am Di., 21. Jan. 2020, 11:34:
> […](#)
> Is there any progress in this PR? — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub <#1318?email_source=notifications&email_token=AADOXUDEPINIBRSUIOLSXV3Q63FU3A5CNFSM4JBVJ6I2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJPIQZQ#issuecomment-576620646>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AADOXUHGE7QI4XL335YSUP3Q63FU3ANCNFSM4JBVJ6IQ> .

Update branch with master and thats it

NobodysNightmare commented:
> Update branch with master and thats it

Easier said than done. I now rebased onto master. Essential conflicts I solved:

* Adding the `created_since` scope is now done in the new mixin for access tokens
* I dropped the `order` statement from my change to `authorized_tokens_for`, since it was dropped from that method upstream as well
* I kept the where clause in my change to `authorized_tokens_for`, because it was added equally upstream (this was irritating me the most, I did not yet check which MR introduced it)

nbulaj commented on lib/doorkeeper/models/access_token_mixin.rb:
```diff
@@ -191,9 +191,22 @@ def find_or_create_for(application, resource_owner_id, scopes, expires_in, use_r
+                       revoked_at: nil)
+
+        if valid_from
+          tokens.created_since(valid_from)
```
Do we have specs for this case? :thinking:

nbulaj replied:
Ping here, I think you missed it :grimacing:

nbulaj commented on lib/doorkeeper/models/access_token_mixin.rb:
```diff
@@ -191,9 +191,22 @@ def find_or_create_for(application, resource_owner_id, scopes, expires_in, use_r
+        # by already filtering out tokens whose TTL has expired.
+        # We can't do that if custom expirations times have been configured, though.
+        unless Doorkeeper.configuration.option_defined?(:custom_access_token_expires_in)
+          valid_from = Doorkeeper.configuration.access_token_expires_in.seconds.ago
```
Also we use UTC everywhere, does this `.ago` respects it?

nbulaj commented:
Hi @NobodysNightmare ! Could you please rebase PR to resolve the conflicts? I'll do other changes myself then. Thanks for your contribution!

nbulaj commented:
Or just let me know and I'll do all myself

nbulaj commented:
I have some plans to release a new version on this week, so I've decided to clone some of the work from this PR to a new one with fixes (currently we can't use optimization with `created_since` API). Check it here #1353

Thanks @NobodysNightmare for awesome work <3

NobodysNightmare commented:
Nice to see this being integrated (even partially).

Do I still need to rebase onto master or should I just close the MR?

nbulaj commented:
@NobodysNightmare you can close this PR as most of the changes from it already in master.

existing token lookup need to be refactored and introduced in a new PR. I posted sample solution that uses time math fo different databases, but I still don't sure about it. 

Btw thanks for your work!
